### PR TITLE
Allow to customize the mapped type name for @InnerField and @Field annotations

### PR DESCRIPTION
**Pull request by reta**
In Spring Data OpenSearch [1], which is built on top of Spring Data Elasticsearch, we are  running into cases where both projects differ in incompatible ways. One of such cases is supported mapping field types (baked by `org.springframework.data.elasticsearch.annotations.FieldType`), for example:

```
Flattened("flattened"), //
```

would have to become 

```
Flat_Object("flat_oblect"), //
```

for OpenSearch. 

This pull request adds optional `mappedName` attribute to `@InnerField` and` @Field` annotations (and model classes respectively), to address the possible divergence, for example:

```
@Field(type = FieldType.Flattened, mappedName = "flat_oblect") String flattenedField;
```

It would require a bit more work for OpenSearch users but not the Elasticsearch ones. The change would be non-breaking (defaults to `""` and `FieldType::getMappedType`) and work seamlessly. 

- [X] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [X] **There is a ticket in the bug tracker for the project in our [issue tracker](https://github.com/spring-projects/spring-data-elasticsearch/issues)**. Add the issue number to the _Closes #issue-number_ line below
- [X] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [X] You submit test cases (unit or integration tests) that back your changes.
- [X] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Closes https://github.com/spring-projects/spring-data-elasticsearch/issues/2942

### CODE REVIEW

File: src/test/java/org/springframework/data/elasticsearch/core/index/MappingBuilderIntegrationTests.java [hunk 1]
```diff
@@ -77,6 +78,12 @@ void cleanup() {
 		operations.indexOps(IndexCoordinates.of(indexNameProvider.getPrefix() + "*")).delete();
 	}
 
+	@Test
+	public void shouldSupportAllTypes() {
```
Comment (reta) [inline src/test/java/org/springframework/data/elasticsearch/core/index/MappingBuilderIntegrationTests.java:82]: This one is not strictly related but `EntityWithAllTypes` was not used in tests at all, fixing that

**Comment by reta** [comment]
@sothawo would really appreciate a review when you have some time, thank you

**Comment by sothawo** [comment]
tomorrow, been busy this week 😁

File: src/main/java/org/springframework/data/elasticsearch/annotations/Field.java [hunk 2]
```diff
@@ -240,4 +241,11 @@
 	 * @since 5.1
 	 */
 	boolean storeEmptyValue() default true;
+	
+	/**
+	 * overrides the mapping field type which otherwise will be taken from corresponding {@link FieldType}
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/annotations/Field.java:246]:
```suggestion
	 * overrides the field type in the mapping which otherwise will be taken from corresponding {@link FieldType}
```
Seems a little clearer for me

File: src/main/java/org/springframework/data/elasticsearch/annotations/InnerField.java [hunk 3]
```diff
@@ -171,4 +172,11 @@
 	 * @since 5.4
 	 */
 	KnnIndexOptions[] knnIndexOptions() default {};
+	
+	/**
+	 * overrides the mapping field type which otherwise will be taken from corresponding {@link FieldType}
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/annotations/InnerField.java:177]: same as above

File: src/main/java/org/springframework/data/elasticsearch/core/index/MappingBuilder.java [hunk 4]
```diff
@@ -210,7 +212,7 @@ private void writeTypeHintMapping(ObjectNode propertiesNode) throws IOException
 		}
 
 		private void mapEntity(ObjectNode objectNode, @Nullable ElasticsearchPersistentEntity<?> entity,
-				boolean isRootObject, String nestedObjectFieldName, boolean nestedOrObjectField, FieldType fieldType,
+				boolean isRootObject, String nestedObjectFieldName, boolean nestedOrObjectField, FieldType fieldType, String mappedName,
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/core/index/MappingBuilder.java:215]:
```suggestion
				boolean isRootObject, String nestedObjectFieldName, boolean nestedOrObjectField, FieldType fieldType, String fieldTypemappedName,
```
might be mistaken as as mapped name of the entity

File: src/main/java/org/springframework/data/elasticsearch/annotations/Field.java [hunk 5]
```diff
@@ -240,4 +241,11 @@
...
+	 *
+	 * @since 5.4
+	 */
+	String mappedName() default "";
```
Comment (sothawo) [inline src/main/java/org/springframework/data/elasticsearch/annotations/Field.java:250]:
```suggestion
	String mappedTypeName() default "";
```
to not confuse with the field name

**Review by sothawo — CHANGES_REQUESTED** [review]
Thanks for the PR, I just think that the `mappedName` should be called `mappedTypeName` so that it will not be confused with the field name, that is mapped as well from the Java property name to the `Field.name` attribute

**Comment by reta** [comment]
> Thanks for the PR, I just think that the `mappedName` should be called `mappedTypeName` so that it will not be confused with the field name, that is mapped as well from the Java property name to the `Field.name` attribute

Thanks a lot for the review @sothawo , comments addressed! Thanks again!
